### ISSUE TITLE
Noncompilable C++ code if method has different explicit namespace than the receiver

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/cxx#460]
As observed in https://github.com/dtolnay/cxx/issues/441#issuecomment-725204700:

```rust
#[cxx::bridge]
mod ffi {
    #[namespace = "shared"]
    struct Color {
        r: u8,
        g: u8,
        b: u8,
    }

    #[namespace = "rust_part"]
    extern "Rust" {
        fn is_white(self: &Color) -> bool;
    }
}
```

Currently this results in generating something like the following, which does not compile.

```rust
namespace shared {
struct Color final {
  uint8_t r;
  uint8_t g;
  uint8_t b;

  bool is_white() const noexcept;
};
} // namespace shared

namespace rust_part {
bool Color::is_white() const noexcept {
  ...
}
} // namespace rust_part
```

```console
cxxbridge/sources/demo/src/main.rs.cc:25:6: error: ‘Color’ has not been declared
   25 | bool Color::is_white() const noexcept {
      |      ^~~~~
```

I wouldn't be opposed to accepting that code and just making it put the member function in the struct's namespace all the time, regardless of namespace attribute on the member function.

But if someone feels strongly this should be an error (with a better message), that would be reasonable too.

**Comment by anforowicz** [comment]
I tried addressing this in https://github.com/anforowicz/cxx/tree/receiver-namespace-fix2.  This isn't yet ready to submit as a PR, but I wanted to share a few notes.  (This is not yet ready to submit as a PR, because it builds on commits that belong to another PR which probably should land first.   FWIW the last 2 commits from the branch linked above are directly relevant to this issue: https://github.com/dtolnay/cxx/commit/1b28c544105decf9a73436eda8860c83d64de1be and https://github.com/dtolnay/cxx/commit/2e9ec6d9d8d98fd46dfaf1125af9174c61ccf22f;  these commits depend on test support commits from another PR - e.g. https://github.com/dtolnay/cxx/commit/adf652922a1c01ae3b294f279415e86bc702ae35).

I wanted to share and describe that fixing this issue in a satisfactory way seems surprisingly difficult:

* I think that adding an error when the method's namespace is different from the receiver's namespace would have been easy.  (Well, there is a risk that this may have broken scenarios that work already.  So far I have only added tests for a "shared" type with a Rust-provided method - maybe other combinations didn't depend on the method's namespace and already worked fine?)
* But using the receiver's namespace instead of the method's namespace turned out to be quite difficult:
    - Mutating namespace stored in `Api::ExternFn` seems difficult:
        - When an `Api::ExternFn` is parsed, we don't yet have information about the namespace of the receiver (because the type declaration may come later in the token stream)
        - `Types::collect` takes an immutable borrow of `&[Api]` which means that we can't mutate `namespace` stored inside `Api::ExternFn` at a later point.
        - We could in theory try mutating `Api::ExternFn` *before* calling `Types::collect`, but that would to some extent mean reimplementing `Types::resolve` which seems like something we would prefer to avoid.
        - We could avoid an immutable borrow of `&[Api]` in an extracted part of `Types` for handling type resolution.  But even `Types::resolutions` have some `'a` references which seem useful.
    - So, we can keep an incorrect namespace value inside `ExternFn`, but use a new `ExternFn::resolve_namespace` method instead of accessing the incorrect value.
        - We can "try to be careful" not to use the incorrect value. This is what my changes above do currently :-/. This seems a bit icky (but maybe okay).
            - At one point I wondered if the incorrect value should be something like `"cxx_bug_should_use_receiver_namespace"`
            - I note that the `namespace` of some other things also shouldn't be used (e.g. we [use default namespace](https://github.com/dtolnay/cxx/blob/68b827689c684e76d7d28374a006c0441380e9ae/syntax/parse.rs#L153) when constructing a `Pair` that represents field names or [function parameter names](https://github.com/dtolnay/cxx/blob/68b827689c684e76d7d28374a006c0441380e9ae/syntax/parse.rs#L643)).  So from that perspective my changes are not necessarily introducing a _new_ problem, but my change are nevertheless more problematic, because obviously a function may have an associated namespace, while a field/parameter never does.
        - We could try to make `ExternFn::name` field private, and force going through accessors (`rust_name()`, `cxx_name()`, `resolved_namespace(types)`, `cxx_fully_qualified_name(types)`).
            - One downside is that this would to some extent duplicate existing code and `impl`s of `Pair`.  Maybe `ExternFn::create_resolved_pair` would be one answer?
        - And kind of similarily we could introduce `ExternFnPair` which could be used as a type of (still public) `ExternFn::name` field, while keeping `ExternFnPair::namespace` private.

FWIW, `efn.name` appears in the following places (this is _after_ already replacing some occurences of `efn.name.namespace` in `mangle.rs` and in `write.rs`;  and this is after filtering out `efn.name.rust` for which a `rust_name` accessor seems reasonable):

```
$ rg '\befn.name\b' | grep -v efn.name.rust
syntax/ident.rs:                check(cx, &efn.name);   // <= reuses `Pair` check-ing
gen/src/namespace.rs:            Api::CxxFunction(efn) | Api::RustFunction(efn) => &efn.name.namespace,  // No `types` here yet.
gen/src/write.rs:        None => write!(out, "{}", efn.name.to_fully_qualified()),  // Hmmm... have to think about this one more.
gen/src/write.rs:            efn.name.cxx,
gen/src/write.rs:        None => efn.name.cxx.to_string(),
gen/src/write.rs:            efn.name.cxx,
gen/src/write.rs:    let main = efn.name.cxx == *"main"
gen/src/write.rs:        && efn.name.namespace == Namespace::ROOT
```
